// react-redux-form/lib/actions/model-actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _endsWith = require('lodash/endsWith');

var _endsWith2 = _interopRequireDefault(_endsWith);

var _icepick = require('icepick');

var _icepick2 = _interopRequireDefault(_icepick);

var _isEqual = require('lodash/isEqual');

var _isEqual2 = _interopRequireDefault(_isEqual);

var _actionTypes = require('../action-types');

var _actionTypes2 = _interopRequireDefault(_actionTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function isEvent(event) {
  return !!(event && event.stopPropagation && event.preventDefault);
}

function getValue(event) {
  return isEvent(event) ? event.target.value : event;
}

function isMulti(model) {
  return (0, _endsWith2.default)(model, '[]');
}

var change = function change(model, value) {
  return {
    type: _actionTypes2.default.CHANGE,
    model: model,
    multi: isMulti(model),
    value: getValue(value)
  };
};

var xor = function xor(model, item) {
  return function (dispatch, getState) {
    var state = (0, _get3.default)(getState(), model, []);
    var stateWithoutItem = state.filter(function (stateItem) {
      return !(0, _isEqual2.default)(stateItem, item);
    });
    var value = state.length === stateWithoutItem.length ? [].concat(_toConsumableArray(state), [item]) : stateWithoutItem;

    dispatch({
      type: _actionTypes2.default.CHANGE,
      model: model,
      value: value
    });
  };
};

var push = function push(model) {
  var item = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];
  return function (dispatch, getState) {
    var collection = (0, _get3.default)(getState(), model);
    var value = [].concat(_toConsumableArray(collection || []), [item]);

    dispatch({
      type: _actionTypes2.default.CHANGE,
      model: model,
      value: value
    });
  };
};

var toggle = function toggle(model) {
  return function (dispatch, getState) {
    var value = !(0, _get3.default)(getState(), model);

    dispatch({
      type: _actionTypes2.default.CHANGE,
      model: model,
      value: value
    });
  };
};

var filter = function filter(model) {
  var iterate = arguments.length <= 1 || arguments[1] === undefined ? function (a) {
    return a;
  } : arguments[1];
  return function (dispatch, getState) {
    var collection = (0, _get3.default)(getState(), model);
    var value = collection.filter(iterate);

    dispatch({
      type: _actionTypes2.default.CHANGE,
      model: model,
      value: value
    });
  };
};

var reset = function reset(model) {
  return {
    type: _actionTypes2.default.RESET,
    model: model
  };
};

var map = function map(model) {
  var iterate = arguments.length <= 1 || arguments[1] === undefined ? function (a) {
    return a;
  } : arguments[1];
  return function (dispatch, getState) {
    var collection = (0, _get3.default)(getState(), model, []);
    var value = collection.map(iterate);

    dispatch({
      type: _actionTypes2.default.CHANGE,
      model: model,
      value: value
    });
  };
};

var remove = function remove(model, index) {
  return function (dispatch, getState) {
    var collection = (0, _get3.default)(getState(), model, []);

    dispatch({
      type: _actionTypes2.default.CHANGE,
      model: model,
      value: _icepick2.default.splice(collection, index, 1)
    });
  };
};

var merge = function merge(model, values) {
  return function (dispatch, getState) {
    var value = (0, _get3.default)(getState(), model, {});

    dispatch({
      type: _actionTypes2.default.CHANGE,
      model: model,
      value: _icepick2.default.merge(value, values)
    });
  };
};

exports.default = {
  change: change,
  filter: filter,
  map: map,
  merge: merge,
  push: push,
  remove: remove,
  reset: reset,
  toggle: toggle,
  xor: xor
};
