// react-redux-form/lib/native/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Field = undefined;

var _fieldComponent = require('../components/field-component');

var NativeField = (0, _fieldComponent.createFieldClass)({
  MapView: function MapView(props) {
    return {
      onRegionChange: props.onChange,
      onRegionChangeComplete: props.onBlur,
      region: props.modelValue
    };
  },
  Picker: function Picker(props) {
    return {
      onValueChange: props.onChange,
      selectedValue: props.modelValue
    };
  },
  Switch: function Switch(props) {
    return {
      onValueChange: props.onChange
    };
  },
  TextInput: function TextInput(props) {
    return {
      defaultValue: props.modelValue,
      onChangeText: props.onChange,
      onBlur: props.onBlur,
      onFocus: props.onFocus
    };
  },
  DatePickerIOS: function DatePickerIOS(props) {
    return {
      date: props.modelValue,
      onDateChange: props.onChange
    };
  },
  PickerIOS: function PickerIOS(props) {
    return {
      onValueChange: props.onChange,
      selectedValue: props.modelValue
    };
  },
  SegmentedControlIOS: function SegmentedControlIOS(props) {
    return {
      onValueChange: props.onChange,
      selectedIndex: props.values.indexOf(props.modelValue)
    };
  },
  SliderIOS: function SliderIOS(props) {
    return {
      onSlidingComplete: props.onBlur,
      onValueChange: props.onChange,
      value: props.modelValue
    };
  }
});

exports.Field = NativeField;
