// react-redux-form/lib/reducers/form-reducer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getField = exports.initialFormState = exports.initialFieldState = exports.createFormReducer = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

var _every = require('lodash/every');

var _every2 = _interopRequireDefault(_every);

var _icepick = require('icepick');

var _icepick2 = _interopRequireDefault(_icepick);

var _isBoolean = require('lodash/isBoolean');

var _isBoolean2 = _interopRequireDefault(_isBoolean);

var _isEqual = require('lodash/isEqual');

var _isEqual2 = _interopRequireDefault(_isEqual);

var _isPlainObject = require('lodash/isPlainObject');

var _isPlainObject2 = _interopRequireDefault(_isPlainObject);

var _mapValues = require('lodash/mapValues');

var _mapValues2 = _interopRequireDefault(_mapValues);

var _toPath = require('lodash/toPath');

var _toPath2 = _interopRequireDefault(_toPath);

var _actionTypes = require('../action-types');

var _actionTypes2 = _interopRequireDefault(_actionTypes);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var initialFieldState = {
  blur: true,
  dirty: false,
  focus: false,
  pending: false,
  pristine: true,
  submitted: false,
  touched: false,
  untouched: true,
  valid: true,
  validating: false,
  viewValue: null,
  validity: {},
  errors: {}
};

var initialFormState = _extends({}, initialFieldState, {
  fields: {}
});

function getField(state, path) {
  if (!(0, _isPlainObject2.default)(state) || !state.fields) {
    throw new Error('Error when trying to retrieve field \'' + path + ' from an invalid/empty form state. Must pass in a valid form state as the first argument.');
  }

  var localPath = (0, _toPath2.default)(path);

  if (!localPath.length) {
    return state;
  }

  return (0, _get3.default)(state, ['fields', localPath.join('.')], initialFieldState);
}

function setField(state, localPath, props) {
  if (!localPath.length) {
    return _icepick2.default.merge(state, props);
  }

  return _icepick2.default.merge(state, {
    fields: _defineProperty({}, localPath.join('.'), _extends({}, getField(state, localPath), props))
  });
}

function resetField(state, localPath) {
  if (!localPath.length) {
    return initialFormState;
  }

  return _icepick2.default.setIn(state, ['fields', localPath.join('.')], initialFieldState);
}

function formIsValid(formState) {
  return (0, _every2.default)((0, _mapValues2.default)(formState.fields, function (field) {
    return field.valid;
  })) && (0, _every2.default)(formState.errors, function (error) {
    return !error;
  });
}

function createInitialFormState(model) {
  return _extends({}, initialFormState, {
    model: model
  });
}

function createFormReducer(model) {
  var modelPath = (0, _toPath2.default)(model);

  return function () {
    var state = arguments.length <= 0 || arguments[0] === undefined ? createInitialFormState(model) : arguments[0];
    var action = arguments[1];

    if (!action.model) {
      return state;
    }

    var path = (0, _toPath2.default)(action.model);

    if (!(0, _isEqual2.default)(path.slice(0, modelPath.length), modelPath)) {
      return state;
    }

    var localPath = path.slice(modelPath.length);
    var errors = undefined;
    var validity = undefined;

    switch (action.type) {
      case _actionTypes2.default.FOCUS:
        return setField(state, localPath, {
          blur: false,
          focus: true
        });

      case _actionTypes2.default.CHANGE:
      case _actionTypes2.default.SET_DIRTY:
        {
          var setDirtyState = _icepick2.default.merge(state, {
            dirty: true,
            pristine: false
          });

          return setField(setDirtyState, localPath, {
            dirty: true,
            pristine: false
          });
        }

      case _actionTypes2.default.BLUR:
      case _actionTypes2.default.SET_TOUCHED:
        return setField(state, localPath, {
          blur: true,
          focus: false,
          touched: true,
          untouched: false
        });

      case _actionTypes2.default.SET_PENDING:
        return setField(state, localPath, {
          pending: action.pending,
          submitted: false
        });

      case _actionTypes2.default.SET_VALIDITY:
        {
          if ((0, _isPlainObject2.default)(action.validity)) {
            validity = _icepick2.default.merge(getField(state, localPath).validity, action.validity);

            errors = _extends({}, getField(state, localPath).errors, (0, _mapValues2.default)(action.validity, function (valid) {
              return !valid;
            }));
          } else {
            validity = action.validity;
            errors = !action.validity;
          }

          var formIsValidState = setField(state, localPath, {
            errors: errors,
            validity: validity,
            valid: (0, _isBoolean2.default)(errors) ? !errors : (0, _every2.default)(errors, function (error) {
              return !error;
            })
          });

          return _icepick2.default.merge(formIsValidState, {
            valid: formIsValid(formIsValidState)
          });
        }

      case _actionTypes2.default.SET_ERRORS:
        {
          if ((0, _isPlainObject2.default)(action.errors)) {
            validity = _extends({}, getField(state, localPath).validity, (0, _mapValues2.default)(action.errors, function (error) {
              return !error;
            }));

            errors = _icepick2.default.merge(getField(state, localPath).errors, action.errors);
          } else {
            validity = !action.errors;
            errors = action.errors;
          }

          var setErrorsState = setField(state, localPath, {
            errors: errors,
            validity: validity,
            valid: (0, _utils.isValid)(validity)
          });

          return _icepick2.default.merge(setErrorsState, {
            valid: formIsValid(setErrorsState)
          });
        }

      case _actionTypes2.default.SET_PRISTINE:
        {
          var formIsPristine = undefined;
          var setPristineState = undefined;

          if (!localPath.length) {
            formIsPristine = true;

            setPristineState = _icepick2.default.merge(state, {
              fields: (0, _mapValues2.default)(state.fields, function (field) {
                return _extends({}, field, {
                  dirty: false,
                  pristine: true
                });
              })
            });
          } else {
            setPristineState = setField(state, localPath, {
              dirty: false,
              pristine: true
            });

            formIsPristine = (0, _every2.default)((0, _mapValues2.default)(setPristineState.fields, function (field) {
              return field.pristine;
            }));
          }

          return _icepick2.default.merge(setPristineState, {
            dirty: !formIsPristine,
            pristine: formIsPristine
          });
        }

      case _actionTypes2.default.SET_UNTOUCHED:
        return setField(state, localPath, {
          touched: false,
          untouched: true
        });

      case _actionTypes2.default.SET_SUBMITTED:
        return setField(state, localPath, {
          pending: false,
          submitted: !!action.submitted
        });

      case _actionTypes2.default.SET_INITIAL:
      case _actionTypes2.default.RESET:
        return resetField(state, localPath);

      case _actionTypes2.default.SET_VIEW_VALUE:
        return setField(state, localPath, {
          viewValue: action.value
        });

      default:
        return state;
    }
  };
}

exports.createFormReducer = createFormReducer;
exports.initialFieldState = initialFieldState;
exports.initialFormState = initialFormState;
exports.getField = getField;
